File: src/cpu.rs
use crate::memory::{self, Memory};
use crate::shell;
use std::{thread, time};

pub struct Register {
    pub reg_val: i32,
}

pub struct RegisterFile {
    pub pc: Register,
    pub ir0: Register,
    pub ir1: Register,
    pub ac: Register,
    pub mar: Register,
    pub mbr: Register,
    pub base: Register,
}

impl Register {
    fn new(reg_val: i32) -> Register {
        Self { reg_val: reg_val }
    }
}

impl RegisterFile {
    fn new() -> RegisterFile {
        Self {
            pc: Register::new(0),
            ir0: Register::new(0),
            ir1: Register::new(0),
            ac: Register::new(0),
            mar: Register::new(0),
            mbr: Register::new(0),
            base: Register::new(0),
        }
    }
}

pub fn cpu_regs_init() -> RegisterFile {
    RegisterFile::new()
}

fn cpu_mem_address(mut regs: RegisterFile, mut mem: Memory, mem_addr: i32) -> i32 {
    regs.mar.reg_val = regs.base.reg_val + mem_addr;
    memory::mem_read(regs, mem);

    return regs.mbr.reg_val;
}

fn cpu_fetch_instruction(mut regs: RegisterFile, mut mem: Memory) {
    regs.ir0.reg_val = cpu_mem_address(regs, mem, regs.pc.reg_val);
    regs.ir1.reg_val = cpu_mem_address(regs, mem, regs.pc.reg_val + 1);
    regs.pc.reg_val += 2;
}

fn cpu_execute_instruction(mut regs: RegisterFile, mut mem: Memory) {
    match regs.ir0.reg_val {
        0 => { /*Program exit, so do nothing */ }
        1 => regs.ac.reg_val = regs.ir1.reg_val,
        2 => regs.ac.reg_val = cpu_mem_address(regs, mem, regs.ir1.reg_val),
        3 => {
            cpu_mem_address(regs, mem, regs.ir1.reg_val);
            regs.ac.reg_val += regs.mbr.reg_val;
        }
        4 => {
            cpu_mem_address(regs, mem, regs.ir1.reg_val);
            regs.ac.reg_val *= regs.mbr.reg_val;
        }
        5 => {
            regs.mbr.reg_val = regs.ac.reg_val;
            regs.mar.reg_val = regs.base.reg_val + regs.ir1.reg_val;
            memory::mem_write(regs, mem);
        }
        6 => {
            if regs.ac.reg_val > 0 {
                regs.pc.reg_val = regs.ir1.reg_val;
            }
        }
        7 => {
            //TODO change to printer
            println!("AC:{}", regs.ac.reg_val);
        }
        8 => {
            if regs.ir1.reg_val <= 0 {
                eprintln!(
                    "Invalid sleep instruction. sleep time: {} must be positive.",
                    regs.ir1.reg_val
                );
            } else {
                let sleep_millis = time::Duration::from_micros(regs.ir1.reg_val as u64);
                thread::sleep(sleep_millis);
            }
        }
        9 => {
            shell::shell_inst(regs, mem, regs.ir1.reg_val);
        }
        _ => {
            panic!(
                "[cpu.rs] (cpu_execute_instruction) : Invalid Instruction: {}. Exiting.",
                regs.ir0.reg_val
            );
        }
    }
}

fn cpu_operation(regs: RegisterFile, mem: Memory, time_quantum: u16) -> i8 {
    for i in 0..time_quantum {
        if regs.ir0.reg_val == 0 {
            return 1;
        }

        cpu_fetch_instruction(regs, mem);
        cpu_execute_instruction(regs, mem);
    }

    return -1;
}

pub fn cpu_reg_dump(regs: RegisterFile) {
    println!("===========================================");
    println!("               Register Dump");
    println!("===========================================");
    println!("Register: Contents");

    println!("BASE: {}", regs.base.reg_val);
    println!("PC: {}", regs.pc.reg_val);
    println!("IR0: {}", regs.ir0.reg_val);
    println!("IR1: {}", regs.ir1.reg_val);
    println!("AC: {}", regs.ac.reg_val);
    println!("MAR: {}", regs.mar.reg_val);
    println!("MBR: {}", regs.mbr.reg_val);
    print!("\n");
}

File: sim-os-c/project2/cpu.c
#define _GNU_SOURCE
#include <stdio.h>
#include <stdbool.h>
#include <unistd.h>
#include <string.h>
#include "computer.h"

Register_File regs = {{0}, {0}, {0}, {0}, {0}, {0}, {0}};
extern unsigned int curr_pid;
extern unsigned int time_quantum;
extern char cpu_print_buf[33];

int cpu_mem_address(int m_addr)
{
    regs.MAR.reg_val = regs.BASE.reg_val + m_addr;
    mem_read();
    return regs.MBR.reg_val;
}

void cpu_fetch_instruction()
{
    regs.IR0.reg_val = cpu_mem_address(regs.PC.reg_val);
    regs.IR1.reg_val = cpu_mem_address(regs.PC.reg_val + 1);
    regs.PC.reg_val += 2;
}

void cpu_execute_instruction()
{
    if (regs.IR0.reg_val == 1)
    {
        regs.AC.reg_val = regs.IR1.reg_val;
    }
    else if (regs.IR0.reg_val == 2)
    {
        regs.AC.reg_val = cpu_mem_address(regs.IR1.reg_val);
    }
    else if (regs.IR0.reg_val == 3)
    {
        cpu_mem_address(regs.IR1.reg_val);
        regs.AC.reg_val += regs.MBR.reg_val;
    }
    else if (regs.IR0.reg_val == 4)
    {
        cpu_mem_address(regs.IR1.reg_val);
        regs.AC.reg_val *= regs.MBR.reg_val;
    }
    else if (regs.IR0.reg_val == 5)
    {
        regs.MBR.reg_val = regs.AC.reg_val;
        regs.MAR.reg_val = regs.BASE.reg_val + regs.IR1.reg_val;
        mem_write();
    }
    else if (regs.IR0.reg_val == 6)
    {
        if (regs.AC.reg_val > 0)
        {
            regs.PC.reg_val = regs.IR1.reg_val;
        }
    }
    else if (regs.IR0.reg_val == 7)
    {
        memset(cpu_print_buf, 0, sizeof(cpu_print_buf));
        snprintf(cpu_print_buf, 33, "AC:%d", regs.AC.reg_val);
        print_print(curr_pid, cpu_print_buf);
    }
    else if (regs.IR0.reg_val == 8)
    {
        usleep(regs.IR1.reg_val);
    }
    else if (regs.IR0.reg_val == 9)
    {
        shell_instruction(regs.IR1.reg_val);
    }
    else if (regs.IR0.reg_val == 0)
    {
        return;
    }
    else
    {
        printf("[cpu.c] (cpu_execute_instruction) : Invalid Instruction: %d. Exiting.\n", regs.IR0.reg_val);
        regs.IR0.reg_val = 0;
        return;
    }
}

int cpu_operation()
{
    for (int i = 0; i < time_quantum; ++i)
    {
        if (regs.IR0.reg_val == 0)
        {
            return 1;
        }

        cpu_fetch_instruction();
        cpu_execute_instruction();
    }

    return -1;
}

void cpu_reg_dump()
{
    printf("===========================================\n");
    printf("               Register Dump\n");
    printf("===========================================\n");
    printf("Register: Contents\n");

    printf("BASE: %d\n", regs.BASE.reg_val);
    printf("PC: %d\n", regs.PC.reg_val);
    printf("IR0: %d\n", regs.IR0.reg_val);
    printf("IR1: %d\n", regs.IR1.reg_val);
    printf("AC: %d\n", regs.AC.reg_val);
    printf("MAR: %d\n", regs.MAR.reg_val);
    printf("MBR: %d\n", regs.MBR.reg_val);
    printf("\n");
}